fdbclient/MutationLogReader.actor.cpp

```diff
@@ -0,0 +1,196 @@
+	                      1; // firstVersion rounded up to the nearest 1M versions, then + 1
+	int startIndex = indexToRead;
+	while (indexToRead < result.size() && keyRefToVersion(result[indexToRead].key, prefixLen) < stopVersion) {
+		++indexToRead;
+	}
+	if (indexToRead < result.size()) {
+		firstVersion = keyRefToVersion(result[indexToRead].key, prefixLen); // the version of result[indexToRead]
+	}
+	return Standalone<RangeResultRef>(
+	    RangeResultRef(result.slice(startIndex, indexToRead), result.more, result.readThrough), result.arena());
+}
+
+void PipelinedReader::startReading(Database cx) {
+	reader = getNext(cx);
+}
+
+Future<Void> PipelinedReader::getNext(Database cx) {
+	return getNext_impl(this, cx);
+}
+
+ACTOR Future<Void> PipelinedReader::getNext_impl(PipelinedReader* self, Database cx) {
+	state Transaction tr(cx);
+
+	state GetRangeLimits limits(GetRangeLimits::ROW_LIMIT_UNLIMITED,
+	                            (g_network->isSimulated() && !g_simulator.speedUpSimulation)
+	                                ? CLIENT_KNOBS->BACKUP_SIMULATED_LIMIT_BYTES
+	                                : CLIENT_KNOBS->BACKUP_GET_RANGE_LIMIT_BYTES);
+
+	state Key begin = versionToKey(self->currentBeginVersion, self->prefix);
+	state Key end = versionToKey(self->endVersion, self->prefix);
+
+	loop {
+		// Get the lock
+		wait(self->readerLimit.take());
+
+		// Read begin to end forever until successful
+		loop {
+			try {
+				tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+				tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+
+				RangeResult kvs = wait(tr.getRange(KeyRangeRef(begin, end), limits));
+
+				// No more results, send end of stream
+				if (!kvs.empty()) {
+					// Send results to the reads stream
+					self->reads.send(
+					    RangeResultBlock{ .result = kvs,
+					                      .firstVersion = keyRefToVersion(kvs.front().key, self->prefix.size()),
+					                      .lastVersion = keyRefToVersion(kvs.back().key, self->prefix.size()),
+					                      .hash = self->hash,
+					                      .prefixLen = self->prefix.size(),
+					                      .indexToRead = 0 });
+				}
+
+				if (!kvs.more) {
+					self->reads.sendError(end_of_stream());
+					return Void();
+				}
+
+				begin = kvs.readThrough.present() ? kvs.readThrough.get() : keyAfter(kvs.back().key);
+
+				break;
+			} catch (Error& e) {
+				if (e.code() == error_code_transaction_too_old) {
+					// We are using this transaction until it's too old and then resetting to a fresh one,
+					// so we don't need to delay.
+					tr.fullReset();
+				} else {
+					wait(tr.onError(e));
+				}
+			}
+		}
+	}
+}
+
+} // namespace mutation_log_reader
+
+ACTOR Future<Void> MutationLogReader::initializePQ(MutationLogReader* self) {
+	state int h;
+	for (h = 0; h < 256; ++h) {
+		try {
+			mutation_log_reader::RangeResultBlock front = waitNext(self->pipelinedReaders[h]->reads.getFuture());
```

Comment:
This will force to wait readers in the order specified by `h`. Is this desirable? Typically we would just wait for any/all to return and insert into the priority queue, i.e., something like below (parallel wait for all):

```
std::vector<Future<mutation_log_reader::RangeResultBlock>> blocks;

for (h = 0; h < 256; h++) {
  blocks.push_back(waitAndForward(self->pipelinedReaders[h]->reads.getFuture()));
}
waitForAllReady(blocks);

for (const auto& f : blocks) {
  self->priorityQueue.push(f.get());
}
```

Reply:
I see that in waitAndForward, it's also calling waitNext(). Also, this seems inevitable as for the very first round, we always need to wait for all 256 readers to finish. And with waitForAllReady, we can't count the number of error_code_end_of_stream. One workaround is letting each reader sends a dummy result first (https://github.com/RenxuanW/foundationdb/commit/ef835e1af005f68cd92f428e8f97b8cdbcefcd6e), which @satherton doesn't like.

Reply:
@jzhou77 There no serial wait here on any actual async operations - in other words, no waiting on network etc.  This loop does not _start_ the underlying range reads, it just waits on them.  They were already started by the `MutationLogReader` constructor.  

Each stream's waitNext() result must be checked to see if it should be placed in the priority queue or counted as finished, so we have to iterate over all 256 stream futures.  As-implemented, this iteration is done exactly once.  In your suggestion, it would be done 3 times - to build a futures vector, again in `waitForAllReady()`, and then again after to check ready/error states of each future.

While the overhead doesn't matter, I think for the sake of readability it is better to have a single loop and try/catch.

Reply:
I'm confused. Isn't `waitNext` blocking here?

Reply:
I don't understand what operation you believe is held up by waiting here.  The `waitNext()` is waiting, one by one, on operations _that have already begun_ before the enclosing loop starts.  All 256 reader loops have already been started.  This loop is simply waiting for each one to send its first result to its output stream or throw an error.